### ISSUE TITLE
No mutation operation access is allowed on remote schema for unauthenticated users after running hasura metadata apply

### ISSUE THREAD

**Issue by Rokamaku** [hasura/graphql-engine#8455]
### Version Information

Server Version: 2.6.0
CLI Version (for CLI related issue): >= 2.5.0

### Environment

OSS
### What is the expected behaviour?

Unauthenticated users should always be able to access mutation operation provided by the remote schema that enabled public insert/update mutation

### Keywords

mutation
remote schema
unauthenticated users
hasura metadata apply

### What is the current behaviour?

After running hasura metadata apply, all public access to mutation operation on any remote schema that enabled public insert/update mutation is resulting in an error `field not found in type: 'mutation_root'` or `mutation not exists`

### How to reproduce the issue?

1. Allow insert/update from unauthenticated users on a remote schema using web console
2. Export metadata to the local Hasura console
3. Reapply metadata from the local Hasura console

### Any possible solutions?

Modify any permission on the same remote schema again using web console.

After comparing the metadata of remote schema before and after I did the above solution, I found the difference is the missing `schema  { query: Query mutation: Mutation }` in the permission definition schema makes it not work.

### Can you identify the location in the source code where the problem exists?

No
### If the bug is confirmed, would you be willing to submit a PR?

I am not familiar with Go or Haskell

**Comment by Rokamaku** [comment]
I found that issue happens not only to the unauthenticated users but also all other roles

**Comment by Rokamaku** [comment]
I found a workaround that renaming the mutation type on the remote schema to something different from "Mutation". Reexport the metadata and see that the schema definition adds more lines `schema { mutation: MUTATION_TYPE_NAME }`. Reapply metadata again and it works

**Comment by avallete** [comment]
I confirm also stumble upon this issue when upgrading from 2.1.0 to 2.6.2

I'm trying the mentioned workaround, hope it'll work for me as well.

**Comment by BenoitRanque** [comment]
Can we confirm/deny that this issue also happens if the server and cli versions are the same? Would also appreciate some sample metadata so we can reproduce the issue.

**Comment by avallete** [comment]
> Can we confirm/deny that this issue also happens if the server and cli versions are the same? Would also appreciate some sample metadata so we can reproduce the issue.

I'm currently working on a MRE, I'll edit this message once it's available.

**Comment by Rokamaku** [comment]
> Can we confirm/deny that this issue also happens if the server and cli versions are the same? Would also appreciate some sample metadata so we can reproduce the issue.

I do confirm that I check this issue with cli and servers are the same version (I test with 2.5.0, 2.6.0, and 2.6.1). This issue also happens to both normal version and cli-migration image as well and happens to all roles except the admin.

**Comment by mdovn** [comment]
I confirm that this one affect us as well. We are using hasura cloud

**Comment by misha-front** [comment]
I confirm this. Our team ran into this problem by trying server version 2.6.0 with cli 2.6.0 and server version 2.7.0 with cli 2.7.0.

We got around the problem by downgrading the cli version to 2.5.0

**Comment by larebsyed** [comment]
yup I can confirm, this is a very annoying bug on hasura cloud

**Comment by codingkarthik** [comment]
Hey @avallete, Thank you for the [MRE](https://github.com/avallete/mre-hasura-8455) but while I was trying to follow the steps I ran into some errors, and then I noticed that the `apollo` folder is empty, probably that is causing the errors. It will be awesome if you could update it to get it working because I have not had any luck getting it to repro it myself :disappointed: .

**Comment by avallete** [comment]
> Hey @avallete, Thank you for the [MRE](https://github.com/avallete/mre-hasura-8455) but while I was trying to follow the steps I ran into some errors, and then I noticed that the `apollo` folder is empty, probably that is causing the errors. It will be awesome if you could update it to get it working because I have not had any luck getting it to repro it myself 😞 .

Yeah, sorry for that. I did use a boilerplate to recreate the apollo server and seems like it did mess up with my repository by adding it as a submodule instead of adding the content of the folder.

Everything is fixed on the MRE now, you should be able to follow the reproduction steps ;)

Let me know if encounter any new issues.

**Comment by avallete** [comment]
> So, if you have only queries or mutations and they have the default root field name `Query` and `Mutation` respectively, then the following `schema` object is necessary to be included in the permission.
> 
> ```graphql
> schema {
>    query: Query
>    mutation: Mutation
> }
> ```
> 
> @avallete Adding this in your MRE solved the issue :)

Interesting thanks for your solution, looks like we'll be able to implement it easily on the user-land side!
Strange it worked until then though? Also wondering, why does the cli recover after "re-saving" the remote schema permissions, the schema definition of Apollo doesn't change in that case 🤔

**Comment by codingkarthik** [comment]
>  Also wondering, why does the cli recover after "re-saving" the remote schema permissions, the schema definition of Apollo doesn't change in that case thinking

The schema definition does change! When you save the permission through the console, it does add the `schema` object type. You can export the metadata and check it :wink: .

**Comment by codingkarthik** [comment]
Closing this in favor of https://github.com/hasura/graphql-engine/issues/8455#issuecomment-1148541121

**Comment by AThilenius** [comment]
@codingkarthik Sorry, I'm very confused. That linked commend is a workaround, not a solution. This bug persists, I just checked with the latest CLI version `v2.9.0-beta.3`. Metadata export does not add the `schema { ... }` object like it should.

Did I miss something, or is this being marked "won't fix"?

**Comment by maxcan** [comment]
Confirmed this behavior persists in 2.9.0.cli-migrations-v3.

@codingkarthik I'm not sure how you can close an issue in favor of itself.  This should still be open as the workaround is not really suitable for a production environment.

**Comment by dsandip** [comment]
@maxcan @AThilenius Re-opening this till we get more information. cc: @BenoitRanque @birlam

**Comment by maxcan** [comment]
If anyone wants a script to automatically add the schema entry from @AThilenius's workaround, here goes.  Its idempotent and I have it in my CI flow so my `remote_schemas.yaml` file is good for production

```typescript
import { $, argv, fs } from "zx";
import * as yaml from "js-yaml";
async function fixRemoteSchemaYaml() {
  const doc = yaml.load(fs.readFileSync(argv["file"], "utf8"));
  if (!Array.isArray(doc)) {
    throw new Error("remote_schemas.yaml is not an array");
  }
  doc.forEach((remoteSchema) => {
    const perms = remoteSchema.permissions;
    if (!Array.isArray(perms)) {
      throw new Error(
        `remote_schemas.yaml permissions for ${JSON.stringify(
          remoteSchema
        )} is not an array`
      );
    }
    perms.forEach((permObject) => {
      const def = permObject?.definition;
      if (typeof def.schema !== "string") {
        throw new Error(
          `remote_schema permissions for ${JSON.stringify(
            permObject
          )} is not a string`
        );
      }
      if (
        !def.schema.match(/schema +{ +query: +Query +mutation: +Mutation +}/)
      ) {
        def.schema =
          "schema { query: Query mutation: Mutation }\n" + def.schema;
      }
    });
  });
  const output = yaml.dump(doc, { lineWidth: 40 });
  // console.log(
  //   "file: scripts.ts ~ line 118 ~ fixRemoteSchemaYaml ~ output",
  //   output
  // );
  await fs.writeFileSync(argv["file"], output);
}
```

**Comment by aaka3207** [comment]
@maxcan I modified your script a little bit, this one parses the schema using gql-tag, it's just a little safer IMO
```
const yaml = require("js-yaml");
const path = require("path");
const fs = require("fs");
const prettier = require("prettier/standalone");
const plugins = [require("prettier/parser-graphql")];
const gql = require("graphql-tag");

const doc = yaml.load(
  fs.readFileSync(
    path.join(__dirname, "../hasura/metadata/remote_schemas.yaml"),
    "utf8"
  )
);
if (!Array.isArray(doc)) {
  throw new Error("remote_schemas.yaml is not an array");
}
doc.forEach((remoteSchema) => {
  const perms = remoteSchema.permissions;
  if (!Array.isArray(perms)) {
    return;
    throw new Error(
      `remote_schemas.yaml permissions for ${JSON.stringify(
        remoteSchema
      )} is not an array`
    );
  }
  // for each remote schema permission
  perms.forEach((permObject) => {
    let def = permObject.definition;
    if (typeof def.schema !== "string") {
      throw new Error(
        `remote_schema permissions for ${JSON.stringify(
          permObject
        )} is not a string`
      );
    }
    const graphqlObj = gql(def.schema);
    let hasQueryObject = false;
    let hasMutationObject = false;
    let hasSchemaDefintion = false;
    graphqlObj.definitions.forEach((SchemaDef) => {
      console.log(SchemaDef);
      if (SchemaDef.kind === "SchemaDefinition") {
        hasSchemaDefintion = true;
      }
      if (SchemaDef.kind === "ObjectTypeDefinition") {
        if (SchemaDef.name.value === "Query") {
          hasQueryObject = true;
        }
        if (SchemaDef.name.value === "Mutation") {
          hasMutationObject = true;
        }
      }
    });
    if (!hasSchemaDefintion) {
      let newSchemaDef = "schema {";
      if (hasQueryObject) {
        newSchemaDef += " query: Query";
      }
      if (hasMutationObject) {
        newSchemaDef += " mutation: Mutation";
      }
      newSchemaDef += " }\n";
      def.schema = newSchemaDef + def.schema;
    }
    def.schema = prettier.format(def.schema, {
      parser: "graphql",
      plugins,
    });
  });
});
const output = yaml.dump(doc, { lineWidth: 40 });
// console.log(
//   "file: scripts.ts ~ line 118 ~ fixRemoteSchemaYaml ~ output",
//   output
// );
fs.writeFileSync(
  path.join(__dirname, "../hasura/metadata/remote_schemas.yaml"),
  output
);
```

**Comment by maxcan** [comment]
> @maxcan I modified your script a little bit, this one parses the schema using gql-tag, it's just a little safer IMO

That is much better.  Mine was very much a quick hack.

**Comment by BenoitRanque** [comment]
Was able to repro the issue in 2.6.0

Closing as duplicate of #8579 which is now resolved as of the [latest stable release (2.13.0)](https://github.com/hasura/graphql-engine/releases/tag/v2.13.0)

Note this is a CLI issue so you may resolve this problem by updating your CLI without updating your server version, if for any reason you cannot do so (newer CLI versions are generally backwards compatible with older server versions).
